test/hotspot/jtreg/serviceability/jvmti/vthread/ToggleNotifyJvmtiTest/ToggleNotifyJvmtiTest.java

```diff
@@ -0,0 +1,178 @@
+        shouldFinish = true;
+    }
+}
+
+public class ToggleNotifyJvmtiTest {
+    private static final int VTHREADS_CNT = 60;
+    private static final String AGENT_LIB = "ToggleNotifyJvmtiTest";
+    private static final WhiteBox WB = WhiteBox.getWhiteBox();
+
+    private static native boolean IsAgentStarted();
+    private static native int VirtualThreadStartedCount();
+
+    static void log(String str) { System.out.println(str); }
+
+    static public void sleep(long millis) {
+        try {
+            Thread.sleep(millis);
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Interruption in TestedThread.sleep: \n\t" + e);
+        }
+    }
+
+    static TestedThread[] threads = new TestedThread[VTHREADS_CNT];
+    static Thread vts[] = new Thread[VTHREADS_CNT];
+
+    static private synchronized void startThread(int i) {
+        String name = "TestedThread" + i;
+        TestedThread thread = new TestedThread(name);
+        vts[i] = Thread.ofVirtual().name(name).start(thread);
+        thread.ensureReady();
+        threads[i] = thread;
+        log("## Java: started vthread: " + name);
+    }
+
+    static private void startThreads() {
+        log("\n## Java: Starting vthreads");
+        for (int i = 0; i < VTHREADS_CNT; i++) {
+            startThread(i);
+        }
+    }
+
+    static private synchronized void finishThreads() {
+        log("\n## Java: runIt: Finishing vthreads");
+        try {
+            for (int i = 0; i < VTHREADS_CNT; i++) {
+                TestedThread thread = threads[i];
+                if (thread == null) {
+                    break;
+                }
+                thread.letFinish();
+                vts[i].join();
+            }
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    static private void setVirtualThreadsNotifyJvmtiMode(int iter, boolean enable) {
+        sleep(5);
+        WB.setVirtualThreadsNotifyJvmtiMode(enable);
+        log("# main: SetNotifyJvmtiEvents: #" + iter + " enable: " + enable);
+    }
+
+    public static void main(String[] args) throws Exception {
+        log("# main: loading " + AGENT_LIB + " lib");
+
+        if (args.length > 0 && args[0].equals("attach")) { // agent loaded into running VM case
+            String arg = args.length == 2 ? args[1] : "";
+            VirtualMachine vm = VirtualMachine.attach(String.valueOf(ProcessHandle.current().pid()));
+            vm.loadAgentLibrary(AGENT_LIB, arg);
+        } else {
+            System.loadLibrary(AGENT_LIB);
+        }
+        int waitCount = 0;
+        while (!IsAgentStarted()) {
+            log("# main: waiting for native agent to start: #" + waitCount++);
+            sleep(20);
+        }
+        Thread tt = Thread.ofPlatform().name("StartThreadsTest").start(ToggleNotifyJvmtiTest::startThreads);
+        sleep(20);
+
+        for (int iter = 0; VirtualThreadStartedCount() < VTHREADS_CNT; iter++) {
```

Comment:
The test seems to exit once all the threads are started. I would think you would want it to run for a while after all the threads are started.

Reply:
I'm not sure if it is really needed. 60 virtual threads are started.
Some of them are executed long enough before shutdown.
We can just increase the number of threads if necessary.

Reply:
ok

Reply:
I've refactored the test, and it has impacted this fragment. Please, see my comment above.